[ASan][ADT] Don't scribble with ASan

AdvenamTacet opened the pull request:
AddressSanitizer (ASAN) disables scribbling to prevent overwriting poisoned objects.
Needed by https://github.com/llvm/llvm-project/pull/79049

Commit:
[ASan][ADT] Don't scribble with ASan

AddressSanitizer (ASAN) disables scribbling to prevent overwriting poisoned objects.
Needed by https://github.com/llvm/llvm-project/pull/79049

Commit:
clang-format-fix

aeubanks commented:
use `LLVM_ADDRESS_SANITIZER_BUILD` instead from https://github.com/llvm/llvm-project/blob/fe0ec2c91cfbf2aad2a61e402f21b771db685b90/llvm/include/llvm/Support/Compiler.h#L442C10-L442C39

Commit:
Use LLVM_ADDRESS_SANITIZER_BUILD

AdvenamTacet commented:
@aeubanks thx! Just updated my PRs.

aeubanks commented:
code lgtm with comment change

vitalybuka commented on llvm/include/llvm/ADT/FunctionExtras.h:
```diff
@@ -317,8 +317,10 @@ template <typename ReturnT, typename... ParamTs> class UniqueFunctionBase {
 
-#ifndef NDEBUG
-    // In debug builds, we also scribble across the rest of the storage.
+#if !defined(NDEBUG) && !defined(LLVM_ADDRESS_SANITIZER_BUILD)
```
it's always defined, should be 
```
#if !defined(NDEBUG) && !LLVM_ADDRESS_SANITIZER_BUILD
```

vitalybuka replied:
and #include "llvm/Support/Compiler.h" to be sure

AdvenamTacet replied:
Thx, done!

Commit:
Fix based on code review

This commit:
- adds #include "llvm/Support/Compiler.h"
- fixes incorrect use of LLVM_ADDRESS_SANITIZER_BUILD

AdvenamTacet commented:
> code lgtm with comment change

I don't see any comment from you, should I change something more? Let me know and if not, please approve the review.

AdvenamTacet commented:
I ran `buildbot_bootstrap_asan` locally with that change and all ASan tests passed.

vitalybuka commented on llvm/include/llvm/ADT/FunctionExtras.h:
```diff
@@ -317,8 +318,10 @@ template <typename ReturnT, typename... ParamTs> class UniqueFunctionBase {
 
-#ifndef NDEBUG
-    // In debug builds, we also scribble across the rest of the storage.
+#if !defined(NDEBUG) && !LLVM_ADDRESS_SANITIZER_BUILD
```
&& !LLVM_MEMORY_SANITIZER_BUILD # because memset makes uninitialzied stuff initialzed 
&& !LLVM_HWADDRESS_SANITIZER_BUILD   # I hope we will do similar to asan std:: support this year

vitalybuka replied:
Actually unrelated, lets not do in this patch

Commit:
Update comment text.

AdvenamTacet commented on llvm/include/llvm/ADT/FunctionExtras.h:
```diff
@@ -317,8 +317,10 @@ template <typename ReturnT, typename... ParamTs> class UniqueFunctionBase {
-#ifndef NDEBUG
-    // In debug builds, we also scribble across the rest of the storage.
+#if !defined(NDEBUG) && !defined(LLVM_ADDRESS_SANITIZER_BUILD)
+    // In debug builds without ASan, we also scribble across the rest of the
```
I fixed the text to be more precise. Thx!
